File: shared/src/commonMain/kotlin/com/vnidrop/app/feature/receive/ExternalInvitationController.kt
package com.vnidrop.app.feature.receive

import kotlinx.coroutines.channels.Channel
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.receiveAsFlow

const val VniDropInvitationMimeType = "application/vnd.vnidrop.transfer"
const val VniDropInvitationExtension = "vnd"
const val MaxVniDropInvitationBytes = 64 * 1024

/**
 * Buffered ingress for invitation documents opened by a platform host.
 *
 * Hosts can submit before Compose is attached during a cold launch. Each
 * document is then consumed exactly once by the app-level receive workflow.
 */
class ExternalInvitationController {
	// OS document-open dispatch can be triggered by another process. Keep the
	// cold-launch queue bounded so repeated intents cannot grow memory forever.
	private val pending = Channel<Result<String>>(capacity = 16)
	val invitations: Flow<Result<String>> = pending.receiveAsFlow()

	fun openInvitation(raw: String) {
		pending.trySend(validateInvitation(raw))
	}

	fun reportOpenFailure(message: String) {
		pending.trySend(Result.failure(IllegalArgumentException(message)))
	}
}

internal fun validateInvitation(raw: String): Result<String> = runCatching {
	require(raw.isNotBlank()) { "The invitation is empty" }
	require(raw.encodeToByteArray().size <= MaxVniDropInvitationBytes) { "The invitation is too large" }
	raw
}

/**
 * Decode invitation document bytes as strict UTF-8 text.
 *
 * Hosts often receive invitation files as opaque binary streams. Reject payloads
 * that are not valid UTF-8 so binary junk never reaches ticket inspection.
 */
fun decodeInvitationBytes(bytes: ByteArray): String {
	require(bytes.isNotEmpty()) { "The invitation is empty" }
	require(bytes.size <= MaxVniDropInvitationBytes) { "The invitation is too large" }
	val text = bytes.decodeToString()
	// decodeToString() replaces malformed sequences; require a lossless round-trip.
	require(text.encodeToByteArray().contentEquals(bytes)) { "The invitation is not valid text" }
	require(text.isNotBlank()) { "The invitation is empty" }
	return text
}

File: apple/VniDrop/Core/ExternalInvitationController.swift
import Foundation

let vniDropInvitationMimeType = "application/vnd.vnidrop.transfer"
let vniDropInvitationExtension = "vnd"
let maxVniDropInvitationBytes = 64 * 1024

/// Buffered ingress for invitation documents opened by the OS, ported from
/// `ExternalInvitationController.kt`. Hosts can submit before the UI is attached
/// during a cold launch; each document is consumed exactly once.
@MainActor
final class ExternalInvitationController: ObservableObject {
	/// Emits validated (or failed) invitations. The app-level receive workflow
	/// consumes each exactly once.
	private var continuation: AsyncStream<Result<String, Error>>.Continuation?
	lazy var invitations: AsyncStream<Result<String, Error>> = {
		AsyncStream { continuation in
			self.continuation = continuation
		}
	}()

	func openInvitation(raw: String) {
		continuation?.yield(validateInvitation(raw))
	}

	func reportOpenFailure(message: String) {
		continuation?.yield(.failure(InvitationError.message(message)))
	}
}

enum InvitationError: LocalizedError {
	case empty
	case tooLarge
	case invalidEncoding
	case message(String)

	var errorDescription: String? {
		switch self {
		case .empty: return "The invitation is empty"
		case .tooLarge: return "The invitation is too large"
		case .invalidEncoding: return "The invitation is not valid text"
		case .message(let m): return m
		}
	}
}

func validateInvitation(_ raw: String) -> Result<String, Error> {
	if raw.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty {
		return .failure(InvitationError.empty)
	}
	if raw.utf8.count > maxVniDropInvitationBytes {
		return .failure(InvitationError.tooLarge)
	}
	return .success(raw)
}

/// Decode invitation document bytes as strict UTF-8, ported from
/// `decodeInvitationBytes`. Rejects payloads that are not lossless UTF-8.
func decodeInvitationBytes(_ bytes: Data) throws -> String {
	guard !bytes.isEmpty else { throw InvitationError.empty }
	guard bytes.count <= maxVniDropInvitationBytes else { throw InvitationError.tooLarge }
	guard let text = String(data: bytes, encoding: .utf8), Data(text.utf8) == bytes else {
		throw InvitationError.invalidEncoding
	}
	guard !text.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty else { throw InvitationError.empty }
	return text
}
